src/Nox.Generator/Infrastructure/Persistence/DbContext.template.cs

```diff
@@ -108,15 +113,70 @@ public override async Task<int> SaveChangesAsync(CancellationToken cancellationT
         try
         {
             HandleSystemFields();
+            await HandleDomainEvents();
             return await base.SaveChangesAsync(cancellationToken);
-
         }
         catch(DbUpdateConcurrencyException)
         {
             throw new Nox.Exceptions.ConcurrencyException($"Latest value of {nameof(IEntityConcurrent.Etag)} must be provided");
         }
     }
 
+    private async Task HandleDomainEvents()
+    {
+        var domainEvents =  GetPendingDomainEvents();
+        
+        await DispatchEvents(domainEvents);
+        
+        ClearDomainEvents();
+    }
+
+    private List<IDomainEvent> GetPendingDomainEvents()
+    {
+        List<IDomainEvent> domainEvents = new();
+        foreach (var entry in ChangeTracker.Entries<IEntityHaveDomainEvents>())
+        {
+            RaiseDomainEvent(entry);
+            domainEvents.AddRange(entry.Entity.DomainEvents);
+        }
+
+        return domainEvents;
+    }
+
+    private void RaiseDomainEvent(EntityEntry<IEntityHaveDomainEvents> entry)
+    {
+        switch (entry.State)
+        {
+            case EntityState.Added:
+                entry.Entity.RaiseCreateEvent();
+                break;
+
+            case EntityState.Modified:
+                entry.Entity.RaiseUpdateEvent();
+                break;
+
+            case EntityState.Deleted:
+                entry.Entity.RaiseDeleteEvent();
+                break;
+        }
+    }
+
+    private async Task DispatchEvents(List<IDomainEvent> domainEvents)
```

Comment:
we can improve this was and publish in parallel, something like this:

```
 var tasks = domainEvents.Select(domainEvent => _publisher.Publish(domainEvent));
 await Task.WhenAll(tasks);
```

Reply:
Desired changes were applied